### PR TITLE
feat(review): adversarially stress-test newly-added parsers/regex/validators for their own failure modes (#123)

### LINKED ISSUE #123: feat(review): adversarially stress-test newly-added parsers/regex/validators for their own failure modes
### Problem Statement

When a PR diff introduces parsing/regex/validation/heuristic code, the review pipeline grades it the same way it grades a null check: by re-reading the lines for happy-path correctness. It never treats the new code as a **function whose decision boundary must be characterized**. As a result the bot's own three review rounds on PR #114 (which added `CHAINED_CALL`, `descriptionCitesAbsentCode`, `compact()`/`appearsIn`, and reused the #55 diff-window scope in `FindingQuoteValidator.java`) all returned "No issues found," while human review found ~10 precision/recall defects in that very code — three of which are filed as bugs alongside this issue.

The class is unreachable by construction. Confirming any of those defects requires **inventing an input that is not in the diff**: a lambda/nested-paren argument to break `CHAINED_CALL`'s possessive `\([^()]*+\)` (`:64-66`), an NBSP/line-wrapped citation to break the ASCII-only `WHITESPACE` compaction (`:68`, `:219-221`), an off-diff-but-real caller to expose the `descriptionCitesAbsentCode` demotion (`:161-174`). Both `PrReviewPrompts.USER` and `FindingVerifierPrompts.USER` feed only the diff text — a model reading a regex literal and a diff has no instruction or mechanism to synthesize the string the regex mishandles. A grep for `adversarial|fuzz|failure.mode|counterexample|recall` across `src/main` and the prompts returns nothing.

### Proposed Solution

Add a review capability that, when a diff introduces parsing/regex/validation/heuristic code (a new `Pattern`/`Pattern.compile`, a tokenizer/segmenter, a normalize/compact step, a matching/scoping rule, a threshold/tolerance constant like the 3-line ambiguity window), flips the reviewer into a dedicated **"characterize this heuristic's failure modes"** mode for those hunks, separate from generic correctness reading. In that mode the reviewer:

1. Detects the trigger (regex literal, `Pattern.compile`, manual string/index scanning, whitespace/Unicode normalization, a diff-window/scope rule).
2. Adversarially **synthesizes candidate edge-case inputs** for the new logic — e.g. a lambda body for a paren-counting regex, an NBSP/zero-width/line-wrapped string for a whitespace matcher, an input that lives just outside the scope window for a diff-window heuristic.
3. Reasons explicitly about the **false-positive AND false-negative** behavior of the new code on those inputs, reporting where it silently misses or mis-segments — not just where a visible line is wrong.

This is the missing angle that would have caught all three filed bugs on PR #114. Pair it with #96 (execution oracle): the new capability generates the inputs, #96 compiles and runs the new parser/regex against them to turn an unverifiable "might miss X" into a demonstrated pass/fail (#96 today only runs the diff's *existing* touched tests, so it must be extended to accept synthesized inputs).

### Alternatives Considered

- **#63 (multi-agent review)** parallelizes the same reading angles; it is the natural *host* for this as one specialized agent role, but on its own it does not add adversarial-input synthesis — every current angle would inherit the diff-only blind spot. Recommend implementing this capability and noting #63 as where the agent role lives.
- **#113 (eval/regression corpus)** is a fixed historical replay measuring precision on known past false-positives; it can *store* the PR #114 limitations as labeled recall cases (and should), but it does not *generate* novel adversarial inputs per-PR. Seeding #113 with these cases is an extension of #113, not this.
- **#117 (producer->consumer tracing)** traces a data structure end-to-end — a different mismatch class.
- **#96 (execution oracle)** runs existing tests, it does not synthesize new ones — complementary, see above.

A keyword scan over the relevant open issues (fuzz/adversarial/regex/parser/heuristic/edge-case/recall) returned nothing matching; confirmed genuinely new. Nearest existing numbers: #63 (host), #113 (corpus), #96 (execution).

### Priority

Important

### Additional Context

Three forces compound to make this class invisible, all grounded in the current pipeline:
- **Diff-only, same-modality verification.** `FindingVerificationService.verify` hands the verifier the same escaped diff; `FindingVerifierPrompts.SYSTEM` (`:23-24`) is an even more aggressive false-positive filter ("skeptical senior engineer... eliminate false positives"). A second pass over the same diff is another draw from the same distribution, not a new angle.
- **Conservative tuning deletes the exact register these findings need.** `PrReviewPrompts.java:67` ("If you are uncertain whether an issue is real at all, omit it rather than guess"), `:54` (low "rarely worth reporting — prefer omitting"), and `FindingVerificationService.java:58` (the `\b(may|might|could|potentially|possibly)\b` hedging demotion) all scrub the inherently-probabilistic "this regex MIGHT miss a lambda" finding. The recall-exemption / report-to-summary nuance for heuristic-limitation findings belongs on #105.
- **Quote-grounding forbids reasoning about out-of-diff inputs.** The verbatim-in-diff requirement is correct for hallucinated bugs but structurally rejects a finding about an input that is out-of-diff by definition; the off-diff-caller half is the #55 diff-window limitation now inherited by the new check.

Motivating dogfood case: PR #114 / commit `63755ef`, file `src/main/java/dev/thiagogonzaga/thrillhousebot/review/FindingQuoteValidator.java`.

### Code of Conduct

- [x] I have searched for existing issues and this is not a duplicate.

### PR DESCRIPTION
**Pull request by devops-thiago**
## What type of PR is this?

- [x] ✨ Feature

## Description

When a diff introduces parsing / regex / validation / heuristic code, the pipeline graded it the way
it grades a null check — re-reading the added lines for happy-path correctness. It never treated the
new code as **a function whose decision boundary must be characterized**. Confirming a defect there
requires *inventing an input the diff does not contain by definition*, and both prompts feed only
diff text, so the bot's own three review rounds on PR #114 returned "No issues found" while human
review found ~10 precision/recall defects in that very code.

Three changes:

**1. A new conditional section, `PrReviewPrompts.HEURISTIC_FAILURE_MODES_REQUEST`.** It instructs the
reviewer to identify each new decision rule, *synthesize* the inputs that probe its edges and name
them literally, and report **both** directions — weighting the false negative, since a rule that
silently misses looks correct on the page while a false positive is visible on reading. The
synthesized input must be labelled as absent from the diff ("input not in the diff: …") so it is
never presented as quoted material. Concrete probes are seeded from the shapes in the issue and its
comments: a paren-counting regex meeting a lambda, an ASCII whitespace class meeting an NBSP or
zero-width joiner, a scope window meeting an occurrence one line outside it, a "starts with" check
meeting the token inside a fenced code block (the #160 `/pause` case), and a presence check meeting a
present-but-wrong-shape value (the #157 non-numeric `GITHUB_APP_ID` case).

**2. `HeuristicCodeDetector`** gates it, following `bugFixEfficacySection`'s precedent — the section
is empty unless the diff actually adds such code. Signals: regex construction (Java/JS/Python/Go/
Kotlin), char-level scanning, Unicode/whitespace normalization, a *declared* parse/validate/tokenize
member, and window/threshold constants.

**3. A verifier exemption in `FindingVerifierPrompts.SYSTEM`.** This is the load-bearing part. A
heuristic-limitation finding is *about* an input that cannot appear in the diff, so the existing
verbatim-quote and "unverifiable here" rejection grounds would erase the entire class this capability
generates. The verifier is now told to judge such findings on the rule's mechanics — trace the quoted
rule against the named input — and that inability to execute the rule is not grounds for rejection.

**Deliberately out of scope:** executing the synthesized inputs. That is #96 (unscheduled), which the
issue notes must be extended to accept synthesized inputs rather than only running the diff's
existing tests. This PR delivers generation and reporting, not proof — so findings land at
confidence "low"/"medium" as verification requests.

### Precision over recall on the trigger

The detector deliberately excludes `substring`, `indexOf`, `trim`, `matches` and `parseInt`. They
appear in most diffs, and a trigger that fires on most diffs spends prompt budget on every review
without adding an angle. `parse`/`validate`-style names count only as **declarations**: a declaration keyword on the
same line, no `=` before the name, and no `.` or word char immediately before it — so neither a bare
`parseInt(...)` call nor `private int id = Integer.parseInt(raw)` triggers. Import/package lines are
skipped too, so `import java.text.Normalizer;` cannot stand in for normalization logic. Test files are
skipped — a regex there is usually a fixture. `shouldNotTriggerOnOrdinaryStringHandling` pins this.

### Note on the hedging demotion

The issue lists the `\b(may|might|could...)\b` demotion in `FindingVerificationService` as a force
that scrubs these inherently-probabilistic findings, deferring the fix to #105. That turned out not
to need changing: the demotion drops a hedged blocking finding to *medium confidence*, not to
dropped — it still posts. So no change there, keeping this diff focused.

## Related Issues

Fixes #123

Motivating dogfood case: PR #114 / `FindingQuoteValidator`. Recall seeds named in the issue comments:
#160 (`/pause` inside a code fence), #157 (non-numeric `GITHUB_APP_ID`) — both good candidates to add
to #113's corpus as a follow-up. #63 remains the natural host if this becomes a dedicated agent role.

## How Has This Been Tested?

- [x] Unit tests

- **`HeuristicCodeDetectorTest`** (13 tests, new) — triggers for each signal class in both directions,
  including the two recall seeds, test-file exclusion, removed-heuristic code, and a deleted-file
  hunk.
- **`ReviewPromptAssemblerTest`** (+2) — section empty on ordinary string handling, emitted on a new
  regex.
- **`PrReviewPromptsContentTest`** (+2) — anchors on the request text and the verifier exemption.

Local gates, all green:

- `./mvnw spotless:apply` + `spotless:check`
- `./mvnw verify` — **1800 tests, 0 failures**, SpotBugs `BugInstance size is 0`, JaCoCo gate met
- JaCoCo on the new class: **fully covered**, 0 missed instructions/branches/lines

Coverage-chasing on the last uncovered branch surfaced a real defect, now fixed and pinned by
`shouldNotCarryTheTestFileFlagPastADeletedFile`: a deletion emits `+++ /dev/null`, which the header
regex does not match, so `inTestFile` carried over from the previous file. All `+++ ` lines are now
consumed in one place, which resets the flag and removes a redundant guard.

## Checklist

- [x] My code follows the project's coding standards
- [x] I have performed a self-review of my own code

### CODE REVIEW

File: src/main/java/dev/thiagogonzaga/thrillhousebot/review/HeuristicCodeDetector.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  private static final Pattern NORMALIZATION =
+      Pattern.compile("(?i)\\bNormalizer\\b|isWhitespace|\\bNBSP\\b|zero.?width");
+
+  /**
+   * A <em>declared</em> parse/validate/tokenize-style member — the decision boundary lives in its
+   * body. Requires a declaration keyword on the same line so a call like {@code parseInt(...)} does
+   * not trigger.
+   */
+  private static final Pattern HEURISTIC_DECLARATION =
+      Pattern.compile(
```
[the remark this answers is not in the document]
Comment (devops-thiago) [reply src/main/java/dev/thiagogonzaga/thrillhousebot/review/HeuristicCodeDetector.java:55]: Confirmed and fixed in 0c9dd47 — the gap `[^(\r\n]*` spanned ` int id = Integer.`, so a declaration keyword plus `Integer.parseInt(` did match. The gap now stops at `=` and a `(?<![.\w])` lookbehind rejects a name preceded by a dot or word char. My `shouldNotTriggerOnOrdinaryStringHandling` case used no declaration keyword, which is exactly why it missed this; `shouldNotTriggerOnReceiverCallsThatMerelyStartWithAHeuristicName` now covers `parseInt`/`parseLong`/`validator.validate`. The PR description claim is now accurate rather than aspirational.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/main/java/dev/thiagogonzaga/thrillhousebot/review/HeuristicCodeDetector.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    return false;
+  }
+
+  private static boolean isHeuristicLine(String addedLine) {
+    return REGEX_CONSTRUCTION.matcher(addedLine).find()
+        || CHAR_SCANNING.matcher(addedLine).find()
+        || NORMALIZATION.matcher(addedLine).find()
+        || HEURISTIC_DECLARATION.matcher(addedLine).find()
+        || THRESHOLD_CONSTANT.matcher(addedLine).find();
+  }
```
[the remark this answers is not in the document]
Comment (devops-thiago) [reply src/main/java/dev/thiagogonzaga/thrillhousebot/review/HeuristicCodeDetector.java:134]: Confirmed and fixed in 0c9dd47 — `import java.text.Normalizer;` did trip the normalization signal. I implemented the guard differently than suggested: `addedLine.matches("^\\+\\s*import\\s")` would never fire, because `String.matches()` must match the *entire* line and that pattern ends right after the whitespace. Used a precompiled `DECLARATION_ONLY_LINE` with `find()` instead, also covering `from`/`package`/`using`/`#include`. `shouldStillTriggerWhenTheImportIsAccompaniedByRealUse` pins that an import plus actual use still triggers.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
